src/odemis/gui/cont/project.py

```diff
@@ -178,20 +182,21 @@ def _on_btn_roa(self, _):
 
         # Minimum index that has not yet been deleted, find the first index which is not in the existing indices
         num = next(idx for idx, n in enumerate(sorted(self.roa_ctrls.keys()) + [0], 1) if idx != n)
-        name = "ROA-%s" % num
+        if name is None:
+            name = "ROA-%s" % num
         name = make_unique_name(name, [roa.name.value for roa in self.model.roas.value])
         # better guess for parameters after region is selected in _add_roa_ctrl
         roa_ctrl = FastEMROAController(name, None, None, self.colour, self._tab_data, self.panel, self._viewport)
         self.roa_ctrls[num] = roa_ctrl
         self._current_roa_ctrl = roa_ctrl
 
         roa_ctrl.model.coordinates.subscribe(self._add_roa_ctrl)
+        return roa_ctrl
 
     # already running in main GUI thread as it receives event from GUI
     def _on_btn_remove(self, _, roa_ctrl):
         self.remove_roa_ctrl(roa_ctrl)
 
-    @call_in_wx_main  # call in main thread as changes in GUI are triggered
```

Comment:
`_add_roa_ctrl` is only a one time subscription i.e. for the first coordinates value change it will be called. Running this in main thread makes it blocking, as in if I am importing 3 ROAs in the plugin loop, somehow only the third ROA panel will be created and drawn.

Reply:
Indeed, `call_in_wx_main` means the function is not immediately called. The call directly returns, but the actual function is put on a queue. That's why it's not blocking.

Thera is right, it's quite dangerous. At least, you should add a note in the docstring about the fact that this function should only be called from the main GUI thread. However, as it's a call back to a VA, that's hard to verify.

In addition the `_on_btn_roa()` is supposed to be called when the button is pressed. But now it's also called by your plugin when loading the RoA data. 

I think the proper way to deal with it would be to reorganise the code to have a proper separation between the GUI events, and the creation of a RoA. One approach could be to have a dedicated function to create a new RORController, and a second function to set the coordinates of a RoA. They would expect to run in the main GUI thread. Your plugin can call them. Separately, there would be two functions here to listen to the "Add ROA button" and to listen when the user has defined the coordinates on the first time. They would call each of the raw functions.

Reply:
Thanks Éric for the comment, the restructuring/separation seems to me some work. I have a applied a quick workaround for now, let me know if that will suffice. The change is in `plugins/fastem_roa_import_export.py` from line 164. While trying a few things I found that we cannot have a VA subscribed to `functools.partial`, is that corrrect?

Reply:
Yes, it's better, as it doesn't drop the protection.
> I found that we cannot have a VA subscribed to functools.partial, is that corrrect?

It's "partially" correct (punt intended). You can pass a partial function to subscribe(). However, when you subscribe to a VA, it uses a weak reference. This means that if on the caller side you don't hold a reference to it anymore, the function you pass to the VA will be free'd (like any other Python object). In this case, the VA will not be able to call it. Typically, this is what happens with a partial (or a lambda function) at the end of the scope (ie, the method). To work around it, you'll need to keep a reference to the function you pass to the VA. Usually, we just add this partial function to a private list on the caller class (something like: `self._subscribers.append(my_func)` ).

Reply:
I managed to used `functools.partial` and make it a bit cleaner. Thanks Éric!